Added bld2repo

mcurlej opened the pull request:
Simple tool do download build requirements of a modular build from koji.

@FrostyX @fivaldi can you please take a look? 
I did not modify the spec file. What is the practice here should i add the info to the specfile (changelog etc.) or you will @FrostyX after merge?

Signed-off-by: Martin Curlej <mcurlej@redhat.com>

FrostyX commented on bld2repo/cli.py:
```diff
@@ -0,0 +1,62 @@
+    if args.arch:
+        arch = args.arch
+    else:
+        arch = config.default_arch
```
I would rename the `Config.defaulit_arch` variable to just `Config.arch` and set its value in a constructor. If `None` was passed, `"x86_64"` would be used as default. Then we wouldn't have to calculate the architecture when wanting to use it.

FrostyX commented on bld2repo/test-requirements.txt:
```diff
@@ -0,0 +1,12 @@
+pytest
+
```
So many blank lines :rofl:

FrostyX commented on bld2repo/tests/test_data/pki_core_build.json:
```diff
@@ -0,0 +1,32 @@
+    "volume_id": 9,
+    "volume_name": "rhel-8",
+    "cg_name": null
+}
```
No newline here, and also for the files below. You can see the warnings in the GitHub PR.

FrostyX reviewed (commented):
Thank you very much for the PR @mcurlej!
The tool seems to be very handy.

I made you just a couple of suggestions, nothing major.

> I did not modify the spec file. What is the practice here should i add the info to the specfile (changelog etc.)

Don't change the spec file, the changelog will be generated by `tito` during a release.

mcurlej commented:
@FrostyX 
Thanks for the review, i adjusted the arguments to provide defaults and also display it in help on screen. The `config` receives this information directly from the parser. Please take a look if anything else that needs to be changed.

mcurlej commented:
@FrostyX Hi, so after our discussion on IRC yesterday i updated the spec file and added the newlines to the files. It seems my editor was removing them automatically. Can you take a look?

FrostyX commented:
Thank you for the changes.

There is a problem with building RPM package from this branch because of `generate-manpages.sh` script. Please see
https://copr.fedorainfracloud.org/coprs/frostyx/modulemd-tools/build/2195883/

I think we can solve it like this

```
HERE="$( builtin cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
BASE=$HERE/../

PYTHONPATH=$BASE/bld2repo \
    --pyfile bld2repo/cli.py \
```

Ideally, I would put those variable definitions at the top of the script. What do you think? I am for any other solution too, this was the first thing that crossed my mind.

FrostyX commented:
When I locally fixed the the manpage generation error, it fails on

```
error: Installed (but unpackaged) file(s) found:                                                                                                                                                                                                  /usr/lib/python3.9/site-packages/tests/__init__.py                                                                                                                                                                                             /usr/lib/python3.9/site-packages/tests/__pycache__/__init__.cpython-39.opt-1.pyc                                                                                                                                                               /usr/lib/python3.9/site-packages/tests/__pycache__/__init__.cpython-39.pyc                                                                                                                                                                     /usr/lib/python3.9/site-packages/tests/__pycache__/test_bld2repo.cpython-39.opt-1.pyc                                                                                                                                                          /usr/lib/python3.9/site-packages/tests/__pycache__/test_bld2repo.cpython-39.pyc                                                                                                                                                                /usr/lib/python3.9/site-packages/tests/__pycache__/utils.cpython-39.opt-1.pyc                                                                                                                                                                  /usr/lib/python3.9/site-packages/tests/__pycache__/utils.cpython-39.pyc                                                                                                                                                                        /usr/lib/python3.9/site-packages/tests/test_bld2repo.py                                                                                                                                                                                        /usr/lib/python3.9/site-packages/tests/utils.py
```

FrostyX commented on man/generate-manpages.sh:
```diff
@@ -51,3 +51,13 @@ argparse-manpage \
+    --author-email "mcurlej@redhat.com" \
+    --project-name "bld2repo" \
+    --url HOMEPAGE \
+    > ./man/bld2repo.1
```
And here is one missing newline at the end of file. Your IDE is behaving very strangely IMHO. Basically, the last line in the file should have `\n` at its end (then this warning symbol won't appear in the Github UI) but there shouldn't be an empty line at the end (you won't see a line that is empty but has a line number in the diff).

I don't need it in this PR, just saying so you can fix your IDE settings once you have the time.

kdudka commented:
I needed the following changes to get the package successfully installed on Fedora 33:
```diff
diff --git a/bld2repo/setup.py b/bld2repo/setup.py
index 8112d16..72be06e 100644
--- a/bld2repo/setup.py
+++ b/bld2repo/setup.py
@@ -12,7 +12,7 @@ with open('requirements.txt') as f:
 setup(
     name='bld2repo',
     version='0.1',
-    packages=find_packages(),
+    packages=find_packages(exclude=("tests",)),
     url='https://github.com/rpm-software-management/modulemd-tools',
     license='MIT',
     author='Martin Čurlej',
diff --git a/modulemd-tools.spec b/modulemd-tools.spec
index 9e19851..9568ffd 100644
--- a/modulemd-tools.spec
+++ b/modulemd-tools.spec
@@ -81,7 +81,7 @@ cd bld2repo
 %py3_build
 cd ..

-PYTHONPATH=./modulemd_tools ./man/generate-manpages.sh
+PYTHONPATH=./modulemd_tools:./bld2repo ./man/generate-manpages.sh


 %install
```

Unfortunately, it still tracebacks at run time because `bld2repo/cli.py` is not installed:
```
$ bld2repo
Traceback (most recent call last):
  File "/usr/bin/bld2repo", line 33, in <module>
    sys.exit(load_entry_point('bld2repo==0.1', 'console_scripts', 'bld2repo')())
  File "/usr/bin/bld2repo", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/usr/lib64/python3.9/importlib/metadata.py", line 77, in load
    module = import_module(match.group('module'))
  File "/usr/lib64/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 984, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'cli'
```

mcurlej commented:
@kdudka @FrostyX Thanks for the input guys. 

So the problem was of course between the keyboard and the chair... :D I used entrypoints but the cli code was not part of the package. I debugged it and i was finally able to build it and it seems that bld2repo works now when the package is installed.

Here is the working build https://copr.fedorainfracloud.org/coprs/mcurlej/modulemd-tools/build/2221756/.

kdudka commented:
My bad indeed.  After specifying `--koji-storage-host` the packages are downloaded successfully.  Thanks!

FrostyX commented:
Thank you for the changes.
